README.md
# Pinboard
Package pinboard is a GO library, which uses the Pinboard API to interact 
programatically with your bookmarks, notes and other Pinboard data.

**Note**: This project is not affiliated with [Pinboard.in](https://pinboard.in/about/) in any way.

# Installation

You will need to set up and configure the [Go](https://golang.org/doc/install) 
tool chain.

> go get github.com/umahmood/pinboard

> cd $GOPATH/src/github.com/umahmood/pinboard
    
> go test ./...

# Usage

    import (
        "fmt"
        "github.com/umahmood/pinboard"
    )
    pin := pinboard.New()
    token, err := pin.Auth("username:TOKEN")
    if err := nil {
        ...
    }
    // get all tags in users account.
    tags, err := pin.Tags()
    if err := nil {
        ...
    }
    for _, t := range tags {
        fmt.Println("Name:", t.Name, "# of tagged:", t.Count)
    }

# Documentation

> http://godoc.org/github.com/umahmood/pinboard

# License

See the [LICENSE](LICENSE.md) file for license rights and limitations (MIT).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - 2015-10-28
### Changed
- Ran golint on project resulting in minor code changes.
- Restructured tests into internal and external. internal tests test un-exported 
functionality, external tests test client facing functionality.
- Updated README to clarify installation instructions.

### Added
- Created method IsAuthed to check whether user is authenticated.
- Library is now versioned pinboard.Version().


pinboard1_test.go
package pinboard

import (
	"fmt"
	"net/http"
	"net/http/httptest"
	"net/url"
	"testing"
	"time"
)

// DECODE JSON

func TestDecodeJSONValidDecode(t *testing.T) {
	in := []byte(`{"_id": "55885ac73a66c0a52f732f52",
    "index": "42",
    "guid": "e3356187-5496-4409-a0ff-353a499a30ac",
    "balance": "$3,881.62",
    "picture": "http://placehold.it/32x32",
    "eyeColor": "brown"}`)

	wantData := make(map[string]string)
	wantData["_id"] = "55885ac73a66c0a52f732f52"
	wantData["index"] = "42"
	wantData["guid"] = "e3356187-5496-4409-a0ff-353a499a30ac"
	wantData["balance"] = "$3,881.62"
	wantData["picture"] = "http://placehold.it/32x32"
	wantData["eyeColor"] = "brown"

	var wantError error

	gotData, gotError := decodeJSON(in)

	if gotError != wantError {
		t.Errorf("error: got %v want %v", gotError, wantError)
	}

	wantLen := len(wantData)
	gotLen := len(gotData)

	if wantLen != gotLen {
		t.Errorf("data len: got %d want %d", gotLen, wantLen)
	}

	for k, v := range wantData {
		if gotData[k] != v {
			t.Errorf("data: got %v want %v", gotData, wantData)
		}
	}
}

func TestDecodeJSONInvalidDecode(t *testing.T) {
	// the value 42 is an integer not a string.
	in := []byte(`{"_id": "55885ac73a66c0a52f732f52",
    "index": 42, 
    "guid": "e3356187-5496-4409-a0ff-353a499a30ac",
    "balance": "$3,881.62",
    "picture": "http://placehold.it/32x32",
    "eyeColor": "brown"}`)

	wantError := "json: cannot unmarshal number into Go value of type string"

	gotData, gotError := decodeJSON(in)

	if gotError.Error() != wantError {
		t.Errorf("error: got %s want %s", gotError, wantError)
	}

	if gotData != nil {
		t.Errorf("data: got %v want nil", gotData)
	}
}

func TestDecodeJSONIFaceValidDecode(t *testing.T) {
	in := []byte(`{"count":1,
		"notes":[{"id":"19239102",
				"hash":"b378rhef8herh4f",
				"title":"some note",
				"length":"26",
				"created_at":"2015-04-20 13:51:58",
				"updated_at":"2015-04-20 13:51:58"}]}`)

	gotData, gotError := decodeJSONIFace(in)

	if gotError != nil {
		t.Errorf("error: got %s want nil", gotError)
	}

	if gotData == nil {
		t.Errorf("data: got %v want not nil", gotData)
	}
}

func TestDecodeJSONListIFaceValidDecode(t *testing.T) {
	in := []byte(`[{"popular":["news"]},
		{"recommended":["#socent","internet",
		"iPlayer","news","podcast","socent"]}]`)

	gotData, gotError := decodeJSONListIFace(in)

	if gotError != nil {
		t.Errorf("error: got %v want nil", gotError)
	}

	if gotData == nil {
		t.Errorf("data: got %v want not nil", gotData)
	}
}

func TestDecodeJSONInvalidJSON(t *testing.T) {
	in := []byte(`hello`)

	wantError := "invalid character 'h' looking for beginning of value"

	gotData1, gotError := decodeJSON(in)

	if gotError.Error() != wantError {
		t.Errorf("json error: got %s want %s", gotError, wantError)
	}

	if gotData1 != nil {
		t.Errorf("data: got %v want nil", gotData1)
	}

	gotData2, gotError := decodeJSONIFace(in)

	if gotError.Error() != wantError {
		t.Errorf("json iface error: got %s want %s", gotError, wantError)
	}

	if gotData2 != nil {
		t.Errorf("data: got %v want nil", gotData2)
	}

	gotData3, gotError := decodeJSONListIFace(in)

	if gotError.Error() != wantError {
		t.Errorf("json list iface error: got %s want %s", gotError, wantError)
	}

	if gotData3 != nil {
		t.Errorf("data: got %v want nil", gotData3)
	}
}

func TestDecodeJSONWithEmptyJSON(t *testing.T) {
	in := []byte(``)

	wantError := "unexpected end of JSON input"

	gotData1, gotError := decodeJSON(in)

	if gotError.Error() != wantError {
		t.Errorf("json error: got %s want %s", gotError, wantError)
	}

	if gotData1 != nil {
		t.Errorf("data: got %v want nil", gotData1)
	}

	gotData2, gotError := decodeJSONIFace(in)

	if gotError.Error() != wantError {
		t.Errorf("json iface error: got %s want %s", gotError, wantError)
	}

	if gotData2 != nil {
		t.Errorf("data: got %v want nil", gotData2)
	}

	gotData3, gotError := decodeJSONListIFace(in)

	if gotError.Error() != wantError {
		t.Errorf("json list iface error: got %s want %s", gotError, wantError)
	}

	if gotData3 != nil {
		t.Errorf("data: got %v want nil", gotData3)
	}
}

// PARSEDATETIME

func TestParseDateTimeValid(t *testing.T) {
	in := "2010-02-11 03:46:56"

	want := time.Date(2010, time.February, 11, 3, 46, 56, 0, time.UTC)

	got := parseDateTime(in)

	if got != want {
		t.Errorf("parse date time: got %v want %v", got, want)
	}
}

func TestParseDateTimeInvalid(t *testing.T) {
	in := "2010-02-11"

	want := time.Time{}

	got := parseDateTime(in)

	if got.IsZero() != want.IsZero() {
		t.Errorf("parse date time: got %v want %v", got, want)
	}
}

// STRINGTOBOOL

func TestStringToBool(t *testing.T) {
	got := stringToBool("yes")

	if got != true {
		t.Errorf("string to bool: got %v want true", got)
	}

	got = stringToBool("no")

	if got != false {
		t.Errorf("string to bool: got %v want false", got)
	}
}

// BOOLTOSTRING

func TestBoolToString(t *testing.T) {
	got := boolToString(true)

	if got != "yes" {
		t.Errorf("bool to string: got %v want \"yes\"", got)
	}

	got = boolToString(false)

	if got != "no" {
		t.Errorf("bool to string: got %v want \"no\"", got)
	}
}

// ATOI

func TestAtoi(t *testing.T) {
	got := atoi("42")

	if got != 42 {
		t.Errorf("atoi: got %d want 42", got)
	}

	got = atoi("blah")

	if got != 0 {
		t.Errorf("atoi: got %d want 0", got)
	}
}

// DO

func TestDoForResponseData(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter,
		r *http.Request) {
		fmt.Fprint(w, "{hello:world}")
	}))
	defer ts.Close()

	wantData := []byte(`{hello:world}`)

	var wantError error

	gotData, gotError := do(ts.URL)

	if gotError != wantError {
		t.Errorf("error: got %v want %v", gotError, wantError)
	}

	wantLen := len(wantData)
	gotLen := len(gotData)

	if gotLen != wantLen {
		t.Errorf("data len: got %d want %d", gotLen, wantLen)
	}

	for i := 0; i < wantLen; i++ {
		if gotData[i] != wantData[i] {
			t.Errorf("data: want %v got %v", gotData, wantData)
		}
	}
}

func TestDoWithBadHost(t *testing.T) {
	in := "http://no-such-site-46859755.com"

	wantError := "Get http://no-such-site-46859755.com: dial tcp: lookup " +
		"no-such-site-46859755.com: no such host"

	gotData, gotError := do(in)

	if gotError.Error() != wantError {
		t.Errorf("error: got %s want %s", gotError, wantError)
	}

	if gotData != nil {
		t.Errorf("data: got %v want nil", gotData)
	}
}

func TestDoForServerDown(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter,
		r *http.Request) {
		http.Error(w, "Internal Server Error", http.StatusInternalServerError)
	}))
	defer ts.Close()

	wantError := "HTTP 500 Internal Server Error"

	gotData, gotError := do(ts.URL)

	if gotError.Error() != wantError {
		t.Errorf("error: got %s want %s", gotError, wantError)
	}

	if gotData != nil {
		t.Errorf("data: got %v want nil", gotData)
	}
}

func TestDoFailNilData(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter,
		r *http.Request) {
		http.Error(w, "", http.StatusOK)
	}))
	defer ts.Close()

	wantError := "No data returned from server."

	gotData, gotError := do(ts.URL)

	if gotData != nil {
		t.Errorf("data: got %v want nil", gotData)
	}

	if gotError.Error() != wantError {
		t.Errorf("error: got %v want %v", gotError, wantError)
	}
}

// TOKEN

func TestToken(t *testing.T) {
	p := New()
	p.token = "mango:49349324"

	want := "mango:49349324"

	got := p.Token()

	if got != want {
		t.Errorf("token: got %s want %s", got, want)
	}
}

// MAKEURL

func TestMakeURLWithAuthToken(t *testing.T) {
	inMethod := "posts/all"
	inVals := url.Values{}
	inVals.Set("auth_token", "hello:1234")

	p := New()

	want := "https://api.pinboard.in/v1/posts/all/?auth_token=hello%3A1234&format=json"

	got := p.makeURL(inMethod, inVals)

	if got != want {
		t.Errorf("make url: got %s want %s", got, want)
	}

}

func TestMakeURLWithNilVals(t *testing.T) {
	inMethod := "posts/all"
	p := New()
	p.token = "mango:1234"

	want := "https://api.pinboard.in/v1/posts/all/?auth_token=mango%3A1234&format=json"

	got := p.makeURL(inMethod, nil)

	if got != want {
		t.Errorf("make url: got %s want %s", got, want)
	}
}


doc.go
/*
Package pinboard uses the Pinboard API to interact programatically with your
bookmarks, notes and other Pinboard data.

To start, you must authenticate using the users API token:

    import (
       "github.com/umahmood/pinboard"
    )

    pin := pinboard.New()

    token, err := pin.Auth("username:TOKEN")
    ...

Last time users account had activity:

    t, err := pin.LastUpdate()

    if err != nil {
        ...
    }

    fmt.Println("Last Update:", t)

Add a bookmark:

    b := pinboard.Bookmark{URL: "https://www.eff.org/",
        Title:   "Electronic Frontier Foundation",
        Desc:    "Defending your rights in the digital world",
        Tags:    []string{"privacy", "rights", "encryption"},
        Created: time.Now(),
        Replace: true,
        Shared:  true,
        ToRead:  false,
    }

    ok, err := pin.Add(b)

    if err != nil {
        ...
    }

    if ok {
        fmt.Println("bookmark added!")
    }

Deleting a bookmark:

    ok, err := pin.Del("https://www.eff.org/")

    if err != nil {
        ...
    }

    if ok {
        fmt.Println("bookmark deleted!")
    }

Get the bookmarks added today:

    bmarks, err := pin.Get(time.Time{}, "", nil, false)

    if err != nil {
        ...
    }

    for _, b := range bmarks {
        fmt.Println(b.Title)
    }

Get bookmarks added on April 1st 2015:

    date := time.Date(2015, 4, 1, 0, 0, 0, 0, time.UTC)
    bmarks, err := pin.Get(date, "", nil, false)
    ...

Get bookmarks added on April 1st 2015 which have the tags:

    date := time.Date(2015, 4, 1, 0, 0, 0, 0, time.UTC)
    tags := []string{"python", "maths"}
    bmarks, err := pin.Get(date, "", tags, false)
    ...

Get bookmark for this url:

    bmarks, err := pin.Get(time.Time{}, "https://www.eff.org/", nil, false)
    ...

Number of posts on each date:

    posts, err := pin.Dates(nil)

    if err != nil {
        ...
    }

    for _, p := range posts {
        fmt.Println("Date:", p.Date, "# of posts:", p.Count)
    }

Number of posts on each date for given tags:

    tags := []string{"recipe", "cooking"}

    posts, err := pin.Dates(tags)
    ...

Users 5 most recent posts:

    bmarks, err := pin.Recent(nil, 5)

    if err != nil {
        ...
    }

    for _, b := range bmarks {
        fmt.Println(b.Title)
    }

Users most recent posts filtered by tag:

    tags := []string{"economics"}
    bmarks, err := pin.Recent(tags, 0)
    ...

Get all bookmarks in the users account:

    bmarks, err := pin.Bookmarks(nil, 0, 0, time.Time{}, time.Time{}, false)

    if err != nil {
        ...
    }

    for _, b := range bmarks {
        fmt.Println(b.Title)
    }

Get 10 days worth of bookmarks:

    start := time.Now().AddDate(0, 0, -10)
    end := time.Now()
    bmarks, err := pin.Bookmarks(nil, 0, 0, start, end, false)
    ...

Get all bookmarks which have the tags:

    tags := []string{"ruby", "tutorials"}
    bmarks, err := pin.Bookmarks(tags, 0, 0, time.Time{}, time.Time{}, false)
    ...

Get suggestions for tags based on a URL:

    pop, rec, err := pin.Suggest("https://www.eff.org/")

    if err != nil {
        ...
    }

    fmt.Println("Popular tags:")
    for _, i := range pop {
        fmt.Println(i)
    }

    fmt.Println("Recommended tags:")
    for _, i := range rec {
        fmt.Println(i)
    }

Get a list of all tags in the users account:

    tags, err := pin.Tags()

    if err != nil {
        ...
    }

    for _, t := range tags {
        fmt.Println("Name:", t.Name, "# of tagged:", t.Count)
    }

Delete a tag:

    ok, err := pin.DelTag("fonts")

    if err != nil {
        ...
    }

    if ok {
        fmt.Println("tag deleted.")
    }

Rename a tag:

    ok, err := pin.RenTag("fonts", "typography")

    if err != nil {
        ...
    }

    if ok {
        fmt.Println("tag renamed.")
    }

Get a list of all notes in the users account:

    notes, err := pin.Notes()

    if err != nil {
        ...
    }

    for _, n := range notes {
        fmt.Println("ID", n.ID, "Title", n.Title)
    }

Get an individual note:

    notes, err := pin.Notes()

    if err != nil {
        ...
    }

    for _, m := range notes {
        n, err := pin.NoteID(m.ID)
        if err != nil {
            ...
        }
        fmt.Println(n.Title)
        fmt.Println(n.Text)
    }
*/
package pinboard


pinboard.go
package pinboard

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

// BaseURL points to and specifies the format of the Pinboard API endpoint.
// ** Note ** ideally this URI would be marked as const and only visible inside
// the package. However for testing purposes, the unit tests need to set this URI
// to point to a mock server. ** Do not ** change this URI.
var BaseURL = "https://api.pinboard.in/v1/%s/?%s"

// Pinboard a single instance to interact with bookmarks and other data
type Pinboard struct {
	token  string // e.g. username:TOKEN
	authed bool   // Authenticated with Pinboard service?
}

// Bookmark represents a Pinboard bookmark
type Bookmark struct {
	URL   string
	Title string
	Desc  string
	Tags  []string

	// Creation time for this bookmark.
	Created time.Time
	// Replace any existing bookmark with this URL.
	Replace bool
	// Make bookmark public.
	Shared bool
	// Marks the bookmarks as unread.
	ToRead bool

	Hash []byte // 32 character hexadecimal MD5 hash.
	Meta []byte // 32 character hexadecimal MD5 hash.
}

// Tag represents a Pinboard tag
type Tag struct {
	Name  string
	Count int
}

// Post represents a Pinboard post
type Post struct {
	Date  time.Time
	Count int
}

// NoteMetadata represents Pinboard note meta data
type NoteMetadata struct {
	ID      string
	Title   string
	Length  int       // Size of the note in bytes.
	Hash    []byte    // 20 character long sha1 hash of note text.
	Created time.Time // Time note was created. UTC time.
	Updated time.Time // Time note was last updated. UTC time.
}

// Note represents a Pinboard note
type Note struct {
	NoteMetadata
	Text string
}

// Popular represents Pinboard popular tags
type Popular []string

// Recommended represents Pinboard recommended tags
type Recommended []string

// decodeJSON decodes a JSON structure into map key:string val:string. Decodes
// response data from Auth, LastUpdate, Add, Del, Tags, DelTag, RenTag.
func decodeJSON(jsonBlob []byte) (map[string]string, error) {
	var j map[string]string
	err := json.Unmarshal(jsonBlob, &j)
	if err != nil {
		return nil, err
	}
	return j, nil
}

// decodeJSONIFace decodes a JSON structure into map key:string val:interface{}.
// Decodes response data from Get, Dates, Recent, Notes, NoteID.
func decodeJSONIFace(jsonBlob []byte) (map[string]interface{}, error) {
	var j map[string]interface{}
	err := json.Unmarshal(jsonBlob, &j)
	if err != nil {
		return nil, err
	}
	return j, nil
}

// decodeJSONListIFace decodes a JSON structure into []interface{}. Decodes
// response data from Bookmarks, Suggest.
func decodeJSONListIFace(jsonBlob []byte) ([]interface{}, error) {
	var j []interface{}
	err := json.Unmarshal(jsonBlob, &j)
	if err != nil {
		return nil, err
	}
	return j, nil
}

// parseDateTime parses a date time in the format "2010-02-11 03:46:56" and
// returns a time.Time with location UTC.
func parseDateTime(dt string) time.Time {
	if len(dt) != 19 {
		return time.Time{}
	}
	yr := atoi(dt[0:4])
	mo := atoi(dt[6:7])
	dy := atoi(dt[8:10])
	hr := atoi(dt[11:13])
	mn := atoi(dt[14:16])
	sd := atoi(dt[17:19])
	return time.Date(yr, time.Month(mo), dy, hr, mn, sd, 0, time.UTC)
}

// stringToBool returns true if s == "yes", otherwise false.
func stringToBool(s string) bool {
	if s == "yes" {
		return true
	}
	return false
}

// boolToString return "yes" if b == true, otherwise "no".
func boolToString(b bool) string {
	if b {
		return "yes"
	}
	return "no"
}

// atoi converts a string to an integer, if there is an error in the conversion
// returns 0.
func atoi(s string) int {
	l, err := strconv.Atoi(s)
	if err != nil {
		l = 0
	}
	return l
}

// do performs a HTTP GET on a URL.
func do(url string) (data []byte, err error) {
	rsp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	c := rsp.StatusCode
	if c != http.StatusOK {
		return nil, errors.New("HTTP " + strconv.Itoa(c) + " " + http.StatusText(c))
	}
	defer rsp.Body.Close()
	body, err := ioutil.ReadAll(rsp.Body)
	if err != nil {
		return nil, err
	}
	// server sent back no data or just returned '\n'.
	if body == nil || (len(body) == 1 && body[0] == 10) {
		return nil, errors.New("No data returned from server.")
	}
	return body, nil
}

// New returns a new unauthorized instance of Pinboard.
func New() *Pinboard {
	return &Pinboard{}
}

// Token returns the users token in the format username:TOKEN.
func (p Pinboard) Token() string {
	return p.token
}

// IsAuthed checks if the user has been authenticated with the Pinboard service
func (p Pinboard) IsAuthed() bool {
	return p.authed
}

// makeURL constructs a valid URL which can be used to make a request to the
// Pinboard service.
func (p Pinboard) makeURL(method string, vals url.Values) string {
	if vals == nil {
		vals = url.Values{}
	}
	vals.Set("format", "json")

	a := vals.Get("auth_token")
	if a == "" {
		vals.Set("auth_token", p.token)
	}

	return fmt.Sprintf(BaseURL, method, vals.Encode())
}

// performRequest performs a request to the Pinboard service.
func (p Pinboard) performRequest(method string, vals url.Values) ([]byte, error) {
	// Only calls from Auth(...) can pass when p.auth == false, as we are
	// requesting authentication. For all other calls the API must already be
	// authorized by a previous call to Auth(...).
	if !p.authed && method != "user/api_token" {
		return nil, errors.New("API not authorized")
	}
	url := p.makeURL(method, vals)
	data, err := do(url)
	if err != nil {
		return nil, err
	}
	return data, nil
}

// USER

// Auth validates the provided 'token' with the Pinboard service and returns the
// user's API token, The API token must be in the format: username:TOKEN.
func (p *Pinboard) Auth(token string) (string, error) {
	v := url.Values{}
	v.Set("auth_token", token)
	data, err := p.performRequest("user/api_token", v)
	if err != nil {
		return "", err
	}
	j, err := decodeJSON(data)
	if err != nil {
		return "", err
	}
	p.token = token
	p.authed = true
	return j["result"], nil
}

// POSTS

// LastUpdate returns the last time a bookmark was added, updated or deleted.
// Use this before calling Bookmarks() to see if the data has changed since the
// last fetch.
func (p Pinboard) LastUpdate() (time.Time, error) {
	data, err := p.performRequest("posts/update", nil)
	if err != nil {
		return time.Time{}, err
	}
	j, err := decodeJSON(data)
	if err != nil {
		return time.Time{}, err
	}
	t, err := time.Parse(time.RFC3339, j["update_time"])
	if err != nil {
		return time.Time{}, err
	}
	return t, err
}

// Add adds a bookmark.
func (p *Pinboard) Add(b Bookmark) (bool, error) {
	v := url.Values{}
	// mandatory.
	v.Set("url", b.URL)
	v.Set("description", b.Title)
	// optional
	if b.Desc != "" {
		v.Set("extended", b.Desc)
	}
	if b.Tags != nil {
		var g string
		for _, t := range b.Tags {
			g += t + ","
		}
		v.Set("tags", strings.Trim(g, ","))
	}

	var dt string
	if b.Created.IsZero() {
		// no creation time provided.
		dt = time.Now().UTC().Format(time.RFC3339)
	} else {
		dt = b.Created.UTC().Format(time.RFC3339)
	}
	v.Set("dt", dt)
	v.Set("replace", boolToString(b.Replace))
	v.Set("shared", boolToString(b.Shared))
	v.Set("toread", boolToString(b.ToRead))

	data, err := p.performRequest("posts/add", v)
	if err != nil {
		return false, err
	}
	j, err := decodeJSON(data)
	if err != nil {
		return false, nil
	}
	if j["result_code"] != "done" {
		return false, errors.New(j["result_code"])
	}
	return true, nil
}

// Del deletes a bookmark.
func (p Pinboard) Del(URL string) (bool, error) {
	v := url.Values{}
	v.Set("url", URL)
	data, err := p.performRequest("posts/delete", v)
	if err != nil {
		return false, err
	}
	j, err := decodeJSON(data)
	if err != nil {
		return false, err
	}
	if j["result_code"] != "done" {
		return false, errors.New(j["result_code"])
	}
	return true, nil
}

// Get returns one or more posts on a single day matching the arguments. If no
// date or URL is given, date of most recent bookmark will be used. The meta
// flag allows the ability to include a change detection signature for each
// bookmark.
func (p Pinboard) Get(dt time.Time, URL string, tags []string, meta bool) ([]Bookmark, error) {
	v := url.Values{}
	if !dt.IsZero() {
		v.Set("dt", dt.UTC().Format(time.RFC3339))
	}
	if URL != "" {
		v.Set("url", URL)
	}
	if tags != nil {
		var g string
		for _, t := range tags {
			g += t + ","
		}
		v.Set("tag", strings.Trim(g, ","))
	}
	if meta {
		v.Set("meta", "yes")
	} else {
		v.Set("meta", "no")
	}
	data, err := p.performRequest("posts/get", v)
	if err != nil {
		return nil, err
	}
	j, err := decodeJSONIFace(data)
	if err != nil {
		return nil, err
	}
	var bmarks []Bookmark
	posts := j["posts"].([]interface{})
	for i := 0; i < len(posts); i++ {
		p := posts[i].(map[string]interface{})

		t, err := time.Parse(time.RFC3339, p["time"].(string))
		if err != nil {
			t = time.Time{}
		}
		b := Bookmark{
			URL:     p["href"].(string),
			Title:   p["description"].(string),
			Desc:    p["extended"].(string),
			Tags:    strings.Split(p["tags"].(string), " "),
			Created: t,
			Shared:  stringToBool(p["shared"].(string)),
			ToRead:  stringToBool(p["toread"].(string)),
			Hash:    []byte(p["hash"].(string)),
			Meta:    []byte(p["meta"].(string)),
		}
		bmarks = append(bmarks, b)
	}
	return bmarks, nil
}

// Dates returns a list of dates with the number of posts at each date.
func (p Pinboard) Dates(tags []string) ([]Post, error) {
	v := url.Values{}
	if tags != nil {
		var g string
		for _, t := range tags {
			g += t + ","
		}
		v.Set("tag", strings.Trim(g, ","))
	}
	data, err := p.performRequest("posts/dates", v)
	if err != nil {
		return nil, err
	}
	j, err := decodeJSONIFace(data)
	if err != nil {
		return nil, err
	}
	d := j["dates"].(map[string]interface{})
	var posts []Post
	for k, v := range d {
		w, err := time.Parse("2006-01-02", k)
		if err != nil {
			return nil, err
		}
		posts = append(posts, Post{Date: w, Count: atoi(v.(string))})
	}
	return posts, nil
}

// Recent returns a list of the user's most recent posts, filtered by tag. count
// indicates the number results to return, default is 15 max is 100.
func (p Pinboard) Recent(tags []string, count int) ([]Bookmark, error) {
	v := url.Values{}
	if tags != nil {
		var g string
		for _, t := range tags {
			g += t + ","
		}
		v.Set("tag", strings.Trim(g, ","))
	}
	v.Set("count", strconv.Itoa(count))
	data, err := p.performRequest("posts/recent", v)
	if err != nil {
		return nil, err
	}
	j, err := decodeJSONIFace(data)
	if err != nil {
		return nil, err
	}
	var bmarks []Bookmark
	posts := j["posts"].([]interface{})
	for i := 0; i < len(posts); i++ {
		p := posts[i].(map[string]interface{})

		t, err := time.Parse(time.RFC3339, p["time"].(string))
		if err != nil {
			t = time.Time{}
		}
		b := Bookmark{
			URL:     p["href"].(string),
			Title:   p["description"].(string),
			Desc:    p["extended"].(string),
			Tags:    strings.Split(p["tags"].(string), " "),
			Created: t,
			Shared:  stringToBool(p["shared"].(string)),
			ToRead:  stringToBool(p["toread"].(string)),
			Hash:    []byte(p["hash"].(string)),
			Meta:    []byte(p["meta"].(string)),
		}
		bmarks = append(bmarks, b)
	}
	return bmarks, nil
}

// Bookmarks returns all bookmarks in the user's account. Provides the ability
// to:
// - 'tags' Filter by tags.
// - 'offset' Set offset value.
// - 'count' Set the number of results to return. Default is all.
// - 'start' Return only bookmarks created after this time.
// - 'end' Return only bookmarks created before this time.
// - 'meta' A meta flag to include a change detection signature for each bookmark.
func (p Pinboard) Bookmarks(tags []string, offset int, count int,
	start time.Time, end time.Time, meta bool) ([]Bookmark, error) {

	v := url.Values{}
	if tags != nil {
		var g string
		for _, t := range tags {
			g += t + ","
		}
		v.Set("tag", strings.Trim(g, ","))
	}
	v.Set("start", strconv.Itoa(offset))
	v.Set("results", strconv.Itoa(count))

	if !start.IsZero() {
		v.Set("fromdt", start.UTC().Format(time.RFC3339))
	}
	if !end.IsZero() {
		v.Set("todt", end.UTC().Format(time.RFC3339))
	}
	if meta {
		v.Set("meta", "yes")
	} else {
		v.Set("meta", "no")
	}
	data, err := p.performRequest("posts/all", v)
	if err != nil {
		return nil, err
	}
	j, err := decodeJSONListIFace(data)
	if err != nil {
		return nil, err
	}
	var bmarks []Bookmark
	for i := 0; i < len(j); i++ {
		p := j[i].(map[string]interface{})

		t, err := time.Parse(time.RFC3339, p["time"].(string))
		if err != nil {
			t = time.Time{}
		}
		b := Bookmark{
			URL:     p["href"].(string),
			Title:   p["description"].(string),
			Desc:    p["extended"].(string),
			Tags:    strings.Split(p["tags"].(string), " "),
			Created: t,
			Shared:  stringToBool(p["shared"].(string)),
			ToRead:  stringToBool(p["toread"].(string)),
			Hash:    []byte(p["hash"].(string)),
			Meta:    []byte(p["meta"].(string)),
		}
		bmarks = append(bmarks, b)
	}
	return bmarks, nil
}

// Suggest returns a list of popular tags and recommended tags for a given URL.
// Popular tags are tags used site-wide for the url; recommended tags are drawn
// from the user's own tags.
func (p Pinboard) Suggest(URL string) (Popular, Recommended, error) {
	v := url.Values{}
	v.Set("url", URL)
	data, err := p.performRequest("posts/suggest", v)
	if err != nil {
		return nil, nil, err
	}
	j, err := decodeJSONListIFace(data)
	if err != nil {
		return nil, nil, err
	}
	pop := make(Popular, 0)
	rec := make(Recommended, 0)
	if len(j) >= 2 {
		a := j[0].(map[string]interface{})
		b := a["popular"].([]interface{})
		for _, t := range b {
			pop = append(pop, t.(string))
		}
		a = j[1].(map[string]interface{})
		b = a["recommended"].([]interface{})
		for _, t := range b {
			rec = append(rec, t.(string))
		}
	}
	return pop, rec, nil
}

// TAGS

// Tags returns a full list of the user's tags along with the number of times
// they were used.
func (p Pinboard) Tags() ([]Tag, error) {
	data, err := p.performRequest("tags/get", nil)
	if err != nil {
		return nil, err
	}
	j, err := decodeJSON(data)
	if err != nil {
		return nil, err
	}
	var tags []Tag
	for k, v := range j {
		tags = append(tags, Tag{Name: k, Count: atoi(v)})
	}
	return tags, nil
}

// DelTag delete an existing tag, returns true if the delete operation
// succeeds.
func (p Pinboard) DelTag(tag string) (bool, error) {
	v := url.Values{}
	v.Set("tag", tag)
	data, err := p.performRequest("tags/delete", v)
	if err != nil {
		return false, err
	}
	j, err := decodeJSON(data)
	if err != nil {
		return false, err
	}
	if j["result"] != "done" {
		return false, errors.New(j["result"])
	}
	return true, nil
}

// RenTag rename a tag, or fold it in to an existing tag. Match is not case
// sensitive, returns true if the rename operation succeeds.
func (p Pinboard) RenTag(oldTag, newTag string) (bool, error) {
	v := url.Values{}
	v.Set("old", oldTag)
	v.Set("new", newTag)
	data, err := p.performRequest("tags/rename", v)
	if err != nil {
		return false, err
	}
	j, err := decodeJSON(data)
	if err != nil {
		return false, err
	}
	if j["result"] != "done" {
		return false, errors.New(j["result"])
	}
	return true, nil
}

// NOTES

// Notes returns a list of the user's notes
func (p Pinboard) Notes() ([]NoteMetadata, error) {
	data, err := p.performRequest("notes/list", nil)
	if err != nil {
		return nil, err
	}
	j, err := decodeJSONIFace(data)
	if err != nil {
		return nil, err
	}
	var meta []NoteMetadata
	count := int(j["count"].(float64))
	notes := j["notes"].([]interface{})
	for i := 0; i < count; i++ {
		n := notes[i].(map[string]interface{})
		t := NoteMetadata{ID: n["id"].(string),
			Title:   n["title"].(string),
			Length:  atoi(n["length"].(string)),
			Hash:    []byte(n["hash"].(string)),
			Created: parseDateTime(n["created_at"].(string)),
			Updated: parseDateTime(n["updated_at"].(string)),
		}
		meta = append(meta, t)
	}
	return meta, nil
}

// NoteID given a notes ID, returns an individual user note.
func (p *Pinboard) NoteID(id string) (Note, error) {
	data, err := p.performRequest("notes/"+id, nil)
	if err != nil {
		return Note{}, err
	}
	j, err := decodeJSONIFace(data)
	if err != nil {
		return Note{}, err
	}
	n := Note{}
	n.ID = j["id"].(string)
	n.Title = j["title"].(string)
	n.Length = int(j["length"].(float64))
	n.Hash = []byte(j["hash"].(string))
	n.Created = parseDateTime(j["created_at"].(string))
	n.Updated = parseDateTime(j["updated_at"].(string))
	n.Text = j["text"].(string)
	return n, nil
}
